File: dataset.py
import json
import os
import argparse
from pathlib import Path
from typing import Tuple, List, Dict
import random
from collections import defaultdict

# ==
DEFAULT_DATASET_SIZE = 50
DEFAULT_HOLD_TOLERENCE = 0.035
DEFAULT_OUTPUT_FILE = 'dataset.json'
DEFAULT_SEED = 0
# == 
DATA_PATH = Path('./data/')
STATES_PATH = DATA_PATH / 'states'
LABELS_PATH = DATA_PATH / 'labels'
SCOREBOARD_PATH = DATA_PATH / 'scoreboard'



def collect_data():

    states_files = [os.path.join(STATES_PATH, file) for file in os.listdir(STATES_PATH) if not file.startswith(".")]
    labels_files = [os.path.join(LABELS_PATH, file) for file in os.listdir(LABELS_PATH) if not file.startswith(".")]

    states_files.sort()
    labels_files.sort()

    states = []
    labels = []

    for sf in states_files:
        with open(sf) as f:
            # Lire les lignes
            lines = f.readlines()
            _states = json.loads(lines[0])
            states.append(_states)

    for sf in labels_files:
        with open(sf) as f:
            # Lire les lignes
            lines = f.readlines()
            _labels = json.loads(lines[0])
            labels.append(_labels)
    
    with open(SCOREBOARD_PATH) as f:
        # Lire les lignes
        lines = f.readlines()

        scores = []

        for line in lines:
            line_mod = line.strip().split(':')
            score = int(line_mod[1])
            scores.append(score)

    return states, labels, scores


def write_scoreboard(scoreboard: List[Tuple[str, int]], filename: str):
    with open(filename, 'w') as f:
        for game_id, score in scoreboard:
            f.write(f"{game_id} : {score}\n")


def sort_games(states: List, labels: List, scores: List, reverse: bool=False):

    indexes = list(range(len(scores)))

    indexes.sort(key=scores.__getitem__, reverse=reverse)

    states = map(states.__getitem__, indexes)
    labels = map(labels.__getitem__, indexes)
    scores = map(scores.__getitem__, indexes)

    return states, labels, scores


def get_label_stats_for_game(game_labels : List[Dict[str, int]]):
    stats = defaultdict(float)
    N = 0.
    for label in game_labels:
        stats[list(label.values())[0]] += 1.
        N += 1.
    for key in stats.keys():
        stats[key] /= N

    return stats


def clean_games(states_lists: List, labels_lists: List, scores: List, hold_tolerence: float):

    if hold_tolerence < 0 or hold_tolerence > 1:
        print("[WARNING] hold_tolerence must be a value between [0, 1]")

    clean_states = []
    clean_labels = []
    clean_scores = []
    for game_states, game_labels, game_score in zip(states_lists, labels_lists, scores):
        hold_freq = get_label_stats_for_game(game_labels)[7]
        if not hold_freq > hold_tolerence:
            clean_states.append(game_states)
            clean_labels.append(game_labels)
            clean_scores.append(game_score)
        else:
            print(f'Too many holds: {hold_freq}')

    print(f'Sanity check: best score = {clean_scores[-1]} == 57600')
    print(f'Sanity check: num games = {len(clean_scores)}')

    return clean_states, clean_labels, clean_scores


def select_games(N, states: List, labels: List, ordered_scores: List):
    return states[-N:], labels[-N:], ordered_scores[-N:]


def assemble(states_list: List, labels_list: List):
    states = [state for game_states in states_list for state in game_states]
    labels = [label for game_states in labels_list for label in game_states]

    print(f'Sanity check: num labels = {len(labels)}')
    
    return states, labels

    
def shuffle(states: List, labels: List):
    tuples = list(zip(states,labels))
    random.shuffle(tuples)
    states, labels = zip(*tuples)
    return states, labels
    

def write_dataset(states: List, labels: List, filename: str):
    json_states = json.dumps(states)
    json_labels = json.dumps(labels)

    states_path = DATA_PATH / ('states-' + filename)
    labels_path = DATA_PATH / ('labels-' + filename)
    
    with open(states_path, 'w') as states_file:
        states_file.write(json_states)

    with open(labels_path, 'w') as labels_file:
        labels_file.write(json_labels)

    return states_path, labels_path


def main(args):
    print("Parameters: ")
    print("Seed: \t\t", args.seed)
    print("hold_tolerence:\t", args.hold_tolerence)
    print("shuffle:\t", not args.no_shuffle)
    print("N:\t\t", args.N)
    print("output:\t\t", args.output)

    random.seed(args.seed)

    # Decider limite
    states, labels, scores = collect_data()
    
    # Sort games in descending score order
    states, labels, scores = sort_games(states, labels, scores, reverse=False)
    
    # Clean up
    states, labels, scores = clean_games(states, labels, scores, args.hold_tolerence)

    # Selection des meilleurs games
    states, labels, scores = select_games(args.N, states, labels, scores)

    # Assembler les donnees
    states, labels = assemble(states, labels)
    
    # Shuffle (avec meme seed!)
    if not args.no_shuffle:
        states, labels = shuffle(states, labels)

    # Write to file
    states_path, labels_path = write_dataset(states, labels, args.output)
    print("Dataset successfully written to ", str(states_path), " and ", str(labels_path))

if '__main__' == __name__:
    parser = argparse.ArgumentParser(
        prog='Tetris Dataset Parser',
        description='Parses and selects appropriate training data for Tetris'
    )
    
    parser.add_argument('--N', help='Size of the dataset', default=DEFAULT_DATASET_SIZE, type=int)
    parser.add_argument('--hold_tolerence', help='Tolerence of the number of hold action [0, 1]', type=float, default=DEFAULT_HOLD_TOLERENCE)
    parser.add_argument('--output', help='Output filename in which to export the filtered dataset', type=str, default=DEFAULT_OUTPUT_FILE)
    parser.add_argument('--no_shuffle', help='Disable the shuffling of the states-labels pairs between games.', action='store_false')
    parser.add_argument('--seed', help='Shuffling seed for the states-labels pairs.', type=int, default=DEFAULT_SEED)

    main(parser.parse_args())

File: dataset.jl
using JSON
using Random
using Statistics
using Printf
using ArgParse
using Base.Iterators

const DEFAULT_DATASET_SIZE = 50
const DEFAULT_HOLD_TOLERENCE = 0.035
const DEFAULT_OUTPUT_FILE = "dataset.json"
const DEFAULT_SEED = 0

const DATA_PATH = "data/"
const STATES_PATH = joinpath(DATA_PATH, "states")
const LABELS_PATH = joinpath(DATA_PATH, "labels")
const SCOREBOARD_PATH = joinpath(DATA_PATH, "scoreboard")


function collect_data()
    states_files = [joinpath(STATES_PATH, file) for file in readdir(STATES_PATH) if !startswith(file, ".")]
    labels_files = [joinpath(LABELS_PATH, file) for file in readdir(LABELS_PATH) if !startswith(file, ".")]

    sort!(states_files)
    sort!(labels_files)

    states = Vector{Vector{Dict{String,Vector{Int64}}}}()
    labels = Vector{Vector{Dict{String,Int64}}}()

    for sf in states_files
        _states = JSON.parse(readline(sf), dicttype=Dict{String,Vector{Int64}})
        push!(states, _states)
    end

    for lf in labels_files
        _labels = JSON.parse(readline(lf), dicttype=Dict{String,Int64})
        push!(labels, _labels)
    end

    lines = readlines(SCOREBOARD_PATH)

    scores = Vector{Int64}()
    for line in lines
        score = parse(Int, split(strip(line), ":")[2])
        push!(scores, score)
    end

    return states, labels, scores
end


function write_scoreboard(scoreboard::Vector{Tuple{String,Int}}, filename::String)
    open(filename, "w") do f
        for (game_id, score) in scoreboard
            write(f, "$(game_id) : $(score)\n")
        end
    end
end


function sort_games(states::Vector{Vector{Dict{String,Vector{Int64}}}}, labels::Vector{Vector{Dict{String,Int64}}}, scores::Vector{Int64}, reverse::Bool=false)

    indexes = sortperm(scores, rev=reverse)

    states = states[indexes]
    labels = labels[indexes]
    scores = scores[indexes]

    return states, labels, scores
end


function get_label_stats_for_game(game_labels::Vector{Dict{String,Int64}})
    stats = Dict{Int,Float64}()
    N = 0.
    for i in 2:7
        stats[i] = 0.
    end
    for label in game_labels
        stats[first(values(label))] += 1.
        N += 1.
    end
    for key in keys(stats)
        stats[key] /= N
    end

    return stats
end


function clean_games(states_lists::Vector{Vector{Dict{String,Vector{Int64}}}}, labels_lists::Vector{Vector{Dict{String,Int64}}}, scores::Vector{Int64}, hold_tolerence::Float64)

    if hold_tolerence < 0 || hold_tolerence > 1
        println("[WARNING] hold_tolerence must be a value between [0, 1]")
    end

    clean_states = Vector{Vector{Dict{String,Vector{Int64}}}}()
    clean_labels = Vector{Vector{Dict{String,Int64}}}()
    clean_scores = Int[]

    for (game_states, game_labels, game_score) in zip(states_lists, labels_lists, scores)
        hold_freq = get_label_stats_for_game(game_labels)[7]
        if !(hold_freq > hold_tolerence)
            push!(clean_states, game_states)
            push!(clean_labels, game_labels)
            push!(clean_scores, game_score)
        else
            println("Too many holds: $(hold_freq)")
        end
    end

    println("Sanity check: best score = $(clean_scores[end]) == 57600")
    println("Sanity check: filtered num games = $(length(clean_scores))")

    return clean_states, clean_labels, clean_scores
end

function select_games(N::Int64, states::Vector{Vector{Dict{String,Vector{Int64}}}}, labels::Vector{Vector{Dict{String,Int64}}}, ordered_scores::Vector{Int64})
    return states[end-N+1:end], labels[end-N+1:end], ordered_scores[end-N+1:end]
end

function assemble(states_list::Vector{Vector{Dict{String,Vector{Int64}}}}, labels_list::Vector{Vector{Dict{String,Int64}}})

    println("Sanity check: final num games = ", length(labels_list))

    states = [state for game_states in states_list for state in game_states]
    labels = [label for game_states in labels_list for label in game_states]

    println("Sanity check: num labels = ", length(labels))
    
    return states, labels

end

    
function shuffle(states::Vector{Dict{String,Vector{Int64}}}, labels::Vector{Dict{String,Int64}})
    tuples = collect(zip(states,labels))
    shuffle!(tuples)
    states, labels = unzip(tuples)
    return states, labels
end
    

function write_dataset(states::Vector{Dict{String,Vector{Int64}}}, labels::Vector{Dict{String,Int64}}, filename::AbstractString)
    json_states = JSON.json(states)
    json_labels = JSON.json(labels)

    states_path = joinpath(DATA_PATH, "states-" * filename)
    labels_path = joinpath(DATA_PATH, "labels-" * filename)
    
    open(states_path, "w") do states_file
        write(states_file, json_states)
    end

    open(labels_path, "w") do labels_file
        write(labels_file, json_labels)
    end

    return states_path, labels_path

end

function parse_commandline()
    s = ArgParseSettings()

    @add_arg_table! s begin
        "--N"
            help = "Size of the dataset"
            arg_type = Int
            default = DEFAULT_DATASET_SIZE
        "--hold_tolerence"
            help = "Tolerence of the number of hold action [0, 1]"
            arg_type = Float64
            default = DEFAULT_HOLD_TOLERENCE
        "--output"
            help = "Output filename in which to export the filtered dataset"
            arg_type = String
            default = DEFAULT_OUTPUT_FILE
        "--no_shuffle"
            help = "Disable the shuffling of the states-labels pairs between games."
            action = :store_false
        "--seed"
            help = "Shuffling seed for the states-labels pairs."
            arg_type = Int
            default = DEFAULT_SEED
    end

    return parse_args(s)
end

function main()
    args = parse_commandline()
    println("Parameters: ")
    println("Seed: \t\t", args["seed"])
    println("hold_tolerence:\t", args["hold_tolerence"])
    println("shuffle:\t", !args["no_shuffle"])
    println("N:\t\t", args["N"])
    println("output:\t\t", args["output"])

    Random.seed!(args["seed"])

    # Decider limite
    states, labels, scores = collect_data()
    
    # Sort games in descending score order
    states, labels, scores = sort_games(states, labels, scores)
    
    # Clean up
    states, labels, scores = clean_games(states, labels, scores, args["hold_tolerence"])

    # Selection des meilleurs games
    states, labels, scores = select_games(args["N"], states, labels, scores)

    # Assembler les donnees
    states, labels = assemble(states, labels)
    
    # Shuffle (avec meme seed!)
    if args["no_shuffle"] == false
        states, labels = shuffle(states, labels)
    end

    # Write to file
    states_path, labels_path = write_dataset(states, labels, args["output"])
    println("Dataset successfully written to ", string(states_path), " and ", string(labels_path))

end

main()
   
